File: src/test/scala/pl/msulima/redis/benchmark/ThroughputTest.scala
package pl.msulima.redis.benchmark

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.{CountDownLatch, Executors}

import pl.msulima.redis.benchmark.jedis.JedisClient
import redis.clients.jedis.{Jedis, JedisPool, JedisPoolConfig, Pipeline}

import scala.concurrent.duration._
import scala.concurrent.{Await, ExecutionContext, Future}

object ThroughputTest extends App {

  private val limit = 10000
  private val keys = (1 to limit).map(i => {
    i.toString.getBytes
  }).toArray
  private val items = (1 to limit).map(i => {
    (("." * 100) + i).getBytes
  }).toArray

  private val repeats = 5 * 1000 * 1000
  private val printPeriod = 1000 * 1000

  private implicit val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(9))
  private val pool = new JedisPool(new JedisPoolConfig, "localhost")

  def test(batchSize: Int, repeats: Int): Unit = {
    val start = System.currentTimeMillis()
    val fs = for (j <- 1 to (repeats / batchSize).toInt) yield Future {
      val jedis: Jedis = pool.getResource
      try {
        val pipeline: Pipeline = jedis.pipelined
        var i = 0
        while (i < batchSize) {
          if ((i + j) % 10 == 0) {
            pipeline.set(keys(i), items(i))
          } else {
            pipeline.get(keys(i))
          }
          i = i + 1
        }
        Thread.sleep(1)
        pipeline.sync()
        if (j * batchSize % printPeriod == 0) {
          print(j * batchSize, start)
        }
      } finally {
        if (jedis != null) jedis.close()
      }
    }

    Await.ready(Future.sequence(fs), 10.minutes)
    print(repeats, start)
  }

  def testClient(batchSize: Int, repeats: Int): Unit = {
    val start = System.currentTimeMillis()

    val client = new JedisClient(batchSize)

    var j = 0
    val done = new AtomicInteger()
    val latch = new CountDownLatch(1)

    while (j < repeats) {
      val setCallback = new Runnable {
        override def run(): Unit = {
          val j = done.incrementAndGet()
          if (j == repeats) {
            latch.countDown()
          }
          if (j % printPeriod == 0) {
            print(j, start)
          }
        }
      }
      val getCallback = new java.util.function.Function[Array[Byte], Void] {
        override def apply(t: Array[Byte]): Void = {
          val j = done.incrementAndGet()
          if (j == repeats) {
            latch.countDown()
          }
          if (j % printPeriod == 0) {
            print(j, start)
          }
          null
        }
      }

      val k = j % limit

      if (j % 10 == 0) {
        client.set(keys(k), items(k), setCallback)
      } else {
        client.get(keys(k), getCallback)
      }
      j = j + 1
    }

    latch.await()
    print(repeats, start)
  }

  def suite(test: (Int, Int) => Unit): Unit = {
    test(1000, repeats)
    println("Warmup over")
    Seq(10, 20, 50, 100, 200, 500, 1000, 2000).foreach(batchSize => {
      test(batchSize, repeats)
      println(s"Done $batchSize")
    })
  }

  suite(testClient)
  suite(test)

  System.exit(0)

  def print(repeats: Int, start: Long): Unit = {
    val duration = System.currentTimeMillis() - start
    println((s"${repeats * 1000L / duration} /s", duration))
  }
}

File: src/test/java/pl/msulima/redis/benchmark/jedis/ThroughputTest.java
package pl.msulima.redis.benchmark.jedis;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Timer;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class ThroughputTest {

    private final int setRatio;
    private final byte[][] keys;
    private final byte[][] values;
    private final MetricRegistry metrics;

    public ThroughputTest(int setRatio, byte[][] keys, byte[][] values, MetricRegistry metrics) {
        this.setRatio = setRatio;
        this.keys = keys;
        this.values = values;
        this.metrics = metrics;
    }

    public void run(int batchSize, int repeats) throws InterruptedException {
        Timer meter = metrics.timer(String.format("ThroughputTest(%d, %d)", batchSize, repeats));
        ThroughputTestClient client = new ThroughputTestClient(keys, values, batchSize, meter, setRatio);

        AtomicInteger done = new AtomicInteger();
        CountDownLatch latch = new CountDownLatch(1);

        int throughput = 120_000;
        int sleepTime = 25;
        int perBatch = throughput / (1000 / sleepTime);

        for (int i = 0; i < Math.ceil((double) repeats / perBatch); i++) {
            long start = System.currentTimeMillis();
            for (int j = 0; j < perBatch; j++) {
                client.run(i * perBatch + j, new OnComplete(done, repeats, latch, client));
            }

            long millis = start + sleepTime - System.currentTimeMillis();
            if (millis > 0) {
                Thread.sleep(millis);
            }
        }

        latch.await();
    }

    public static class OnComplete implements Runnable {

        private final AtomicInteger done;
        private final int repeats;
        private final CountDownLatch latch;

        public OnComplete(AtomicInteger done, int repeats, CountDownLatch latch, ThroughputTestClient client) {
            this.done = done;
            this.repeats = repeats;
            this.latch = latch;
        }

        @Override
        public void run() {
            int j = done.incrementAndGet();

            if (j == repeats) {
                latch.countDown();
            }
        }
    }

    public static class ThroughputTestClient {

        private final Timer meter;
        private final int setRatio;
        private final byte[][] keys;
        private final byte[][] values;
        private final JedisClient client;

        public ThroughputTestClient(byte[][] keys, byte[][] values, int batchSize, Timer meter, int setRatio) {
            this.keys = keys;
            this.values = values;
            this.meter = meter;
            this.setRatio = setRatio;
            this.client = new JedisClient(batchSize);
        }

        public void run(int i, Runnable onComplete) {
            long start = System.currentTimeMillis();

            int k = i % keys.length;

            if (i % setRatio == 0) {
                client.set(keys[k], values[k], () -> {
                    meter.update(System.currentTimeMillis() - start, TimeUnit.MILLISECONDS);
                    onComplete.run();
                });
            } else {
                client.get(keys[k], bytes -> {
                    meter.update(System.currentTimeMillis() - start, TimeUnit.MILLISECONDS);
                    onComplete.run();
                    return null;
                });
            }
        }
    }
}
